File: src/geo/surface-engine.ts
// ─── Surface Process Engine ─────────────────────────────────────────────────
// Orchestrator for Phase 3 surface processes: fluvial erosion, glacial
// erosion, aeolian/chemical weathering, and pedogenesis (soil formation).
// Runs after the tectonic engine each tick.

import type { StateBufferViews } from '../shared/types';
import { GRID_SIZE } from '../shared/types';
import type { EventBus } from '../kernel/event-bus';
import type { EventLog } from '../kernel/event-log';
import { ErosionEngine, type ErosionResult } from './erosion-engine';
import { GlacialEngine, type GlacialResult } from './glacial-engine';
import { WeatheringEngine, type WeatheringResult } from './weathering-engine';
import { PedogenesisEngine, type PedogenesisResult } from './pedogenesis';
import type { StratigraphyStack } from './stratigraphy';
import type { Xoshiro256ss } from '../proc/prng';
import { Xoshiro256ss as XoshiroClass } from '../proc/prng';

// ─── Types ──────────────────────────────────────────────────────────────────

export interface SurfaceTickResult {
  erosion: ErosionResult;
  glacial: GlacialResult;
  weathering: WeatheringResult;
  pedogenesis: PedogenesisResult;
}

export interface SurfaceEngineConfig {
  /** Minimum surface process tick interval in Ma (default 0.5). */
  minTickInterval?: number;
  /** Grid size (default GRID_SIZE). */
  gridSize?: number;
}

// ─── SurfaceEngine ──────────────────────────────────────────────────────────

export class SurfaceEngine {
  private readonly erosionEngine: ErosionEngine;
  private readonly glacialEngine: GlacialEngine;
  private readonly weatheringEngine: WeatheringEngine;
  private readonly pedogenesisEngine: PedogenesisEngine;
  private readonly bus: EventBus;
  private readonly eventLog: EventLog;
  private rng: XoshiroClass;

  private stateViews: StateBufferViews | null = null;
  private stratigraphy: StratigraphyStack | null = null;

  /** Time accumulator for sub-tick batching. */
  private accumulator = 0;
  private readonly minTickInterval: number;

  /** Track previous glaciation state for advance/retreat events. */
  private prevGlaciatedCells = 0;

  constructor(
    bus: EventBus,
    eventLog: EventLog,
    seed: number,
    config?: SurfaceEngineConfig,
  ) {
    const gridSize = config?.gridSize ?? GRID_SIZE;
    this.bus = bus;
    this.eventLog = eventLog;
    this.rng = new XoshiroClass(seed);
    this.minTickInterval = config?.minTickInterval ?? 0.5;

    this.erosionEngine = new ErosionEngine(gridSize);
    this.glacialEngine = new GlacialEngine(gridSize);
    this.weatheringEngine = new WeatheringEngine(gridSize);
    this.pedogenesisEngine = new PedogenesisEngine(gridSize);
  }

  /**
   * Initialize the surface engine with shared state and stratigraphy.
   */
  initialize(
    stateViews: StateBufferViews,
    stratigraphy: StratigraphyStack,
  ): void {
    this.stateViews = stateViews;
    this.stratigraphy = stratigraphy;
    this.glacialEngine.clear();
    this.prevGlaciatedCells = 0;
  }

  /**
   * Process one simulation tick. Called after the tectonic engine each frame.
   * Batches sub-ticks if deltaMa exceeds the minimum tick interval.
   */
  tick(timeMa: number, deltaMa: number): SurfaceTickResult | null {
    if (!this.stateViews || !this.stratigraphy || deltaMa <= 0) return null;

    this.accumulator += deltaMa;

    let lastResult: SurfaceTickResult | null = null;

    while (this.accumulator >= this.minTickInterval) {
      const subDelta = this.minTickInterval;
      this.accumulator -= subDelta;
      const subTime = timeMa - this.accumulator;

      lastResult = this.processSurfaceTick(subTime, subDelta);
    }

    return lastResult;
  }

  // ── Core tick processing ──────────────────────────────────────────────

  private processSurfaceTick(timeMa: number, deltaMa: number): SurfaceTickResult {
    const stateViews = this.stateViews!;
    const stratigraphy = this.stratigraphy!;

    // 1 — Fluvial erosion (rivers, sediment transport)
    const erosion = this.erosionEngine.tick(
      timeMa, deltaMa, stateViews, stratigraphy, this.rng,
    );

    // 2 — Glacial erosion (ice sheets, moraines)
    const glacial = this.glacialEngine.tick(
      timeMa, deltaMa, stateViews, stratigraphy, this.rng,
    );

    // 3 — Aeolian & chemical weathering
    const weathering = this.weatheringEngine.tick(
      timeMa, deltaMa, stateViews, stratigraphy, this.rng,
    );

    // 4 — Pedogenesis (soil formation)
    const pedogenesis = this.pedogenesisEngine.tick(
      timeMa, deltaMa, stateViews, stratigraphy,
    );

    // ── Emit events ─────────────────────────────────────────────────────

    // Erosion cycle summary
    if (erosion.cellsAffected > 0) {
      this.bus.emit('EROSION_CYCLE', {
        totalEroded: erosion.totalEroded,
        totalDeposited: erosion.totalDeposited,
        cellsAffected: erosion.cellsAffected,
      });
    }

    // Glaciation advance / retreat
    if (glacial.glaciatedCells > this.prevGlaciatedCells * 1.2 && glacial.glaciatedCells > 10) {
      this.bus.emit('GLACIATION_ADVANCE', {
        glaciatedCells: glacial.glaciatedCells,
        equilibriumLineAltitude: glacial.equilibriumLineAltitude,
      });
      this.eventLog.record({
        timeMa,
        type: 'ICE_AGE_ONSET',
        description: `Glaciation advancing: ${glacial.glaciatedCells} cells glaciated, ELA ${glacial.equilibriumLineAltitude.toFixed(0)} m`,
      });
    } else if (
      glacial.glaciatedCells < this.prevGlaciatedCells * 0.8 &&
      this.prevGlaciatedCells > 10
    ) {
      this.bus.emit('GLACIATION_RETREAT', {
        glaciatedCells: glacial.glaciatedCells,
      });
      this.eventLog.record({
        timeMa,
        type: 'ICE_AGE_END',
        description: `Glaciation retreating: ${glacial.glaciatedCells} cells remaining`,
      });
    }
    this.prevGlaciatedCells = glacial.glaciatedCells;

    // Major river formation
    if (erosion.riverCells.length > 100) {
      const maxAreaCell = erosion.riverCells[0]; // first is often the largest
      this.bus.emit('MAJOR_RIVER_FORMED', {
        cellIndex: maxAreaCell,
        drainageArea: erosion.riverCells.length,
      });
    }

    return { erosion, glacial, weathering, pedogenesis };
  }

  // ── Getters ───────────────────────────────────────────────────────────

  /** Get the glacial engine (for ice thickness visualization). */
  getGlacialEngine(): GlacialEngine {
    return this.glacialEngine;
  }

  /** Get the erosion engine. */
  getErosionEngine(): ErosionEngine {
    return this.erosionEngine;
  }
}

File: backend/GeoTime.Core/Engines/SurfaceEngine.cs
using GeoTime.Core.Compute;
using GeoTime.Core.Models;
using GeoTime.Core.Kernel;
using GeoTime.Core.Proc;

namespace GeoTime.Core.Engines;

/// <summary>Orchestrator for surface processes: erosion, glacial, weathering, pedogenesis.</summary>
public sealed class SurfaceEngine(EventBus bus, EventLog log, uint seed, int gridSize, double minTick = 0.5, GpuComputeService? gpu = null)
{
    private readonly ErosionEngine _erosion = new(gridSize);
    private readonly GlacialEngine _glacial = new(gridSize, gpu);
    private readonly WeatheringEngine _weathering = new(gridSize);
    private readonly PedogenesisEngine _pedogenesis = new(gridSize);
    private readonly Xoshiro256ss _rng = new(seed);

    private SimulationState? _state;
    private StratigraphyStack? _strat;
    private double _accumulator;
    private int _prevGlaciated;

    public void Initialize(SimulationState state, StratigraphyStack strat)
    {
        _state = state; _strat = strat; _glacial.Clear(); _prevGlaciated = 0;
    }

    public SurfaceTickResult? Tick(double timeMa, double deltaMa)
    {
        if (_state == null || _strat == null || deltaMa <= 0) return null;
        _accumulator += deltaMa;
        SurfaceTickResult? last = null;
        while (_accumulator >= minTick)
        {
            _accumulator -= minTick;
            last = Process(timeMa - _accumulator, minTick);
        }
        return last;
    }

    private SurfaceTickResult Process(double t, double dt)
    {
        var er = _erosion.Tick(t, dt, _state!, _strat!, _rng);
        var gl = _glacial.Tick(t, dt, _state!, _strat!, _rng);
        var we = _weathering.Tick(t, dt, _state!, _strat!, _rng);
        var pe = _pedogenesis.Tick(t, dt, _state!, _strat!);

        if (er.CellsAffected > 0)
            bus.Emit("EROSION_CYCLE", new { totalEroded = er.TotalEroded, totalDeposited = er.TotalDeposited, cellsAffected = er.CellsAffected });

        if (gl.GlaciatedCells > _prevGlaciated * 1.2 && gl.GlaciatedCells > 10)
        {
            bus.Emit("GLACIATION_ADVANCE", new { glaciatedCells = gl.GlaciatedCells, equilibriumLineAltitude = gl.EquilibriumLineAltitude });
            log.Record(new GeoLogEntry { TimeMa = t, Type = "ICE_AGE_ONSET", Description = $"Glaciation advancing: {gl.GlaciatedCells} cells" });
        }
        _prevGlaciated = gl.GlaciatedCells;

        return new SurfaceTickResult { Erosion = er, Glacial = gl, Weathering = we, Pedogenesis = pe };
    }

    public GlacialEngine GetGlacialEngine() => _glacial;
}

public sealed class SurfaceTickResult
{
    public ErosionResult Erosion { get; set; } = new();
    public GlacialResult Glacial { get; set; } = new();
    public WeatheringResult Weathering { get; set; } = new();
    public PedogenesisResult Pedogenesis { get; set; } = new();
}
